caches/infinispan/src/main/java/org/commonjava/maven/galley/cache/infinispan/FastLocalCacheProvider.java

```diff
@@ -96,6 +96,12 @@
 
     private PathGenerator pathGenerator;
 
+    // This thread holder is used to add some "re-entrant" like function for the ISPN transaction lock. The ISPN lock is
+    // used for ISPN transaction but not for thread level with re-entrant, that means if we want to own this lock in one
+    // transaction for more than twice in single thread, it will block this thread. So we need this thread holder to by-pass
+    // the ISPN lock waiting when thread not changed.
+    private final ThreadLocal<Long> threadHolder = new ThreadLocal<>();
```

Comment:
Thinking about it more, I don't think we necessarily want ReentrantLock mapped to path. Instead, maybe we just want a set of locked directories local to the thread. ThreadLocal will help us with the "local to the thread" part, but I might prefer using ThreadContext from weft for this, so it will be used across multiple associated threads if we do stream post-processing in parallel.

I'm not sure how this would impact the cluster though. It's locked at the directory level, and that lock owner is shared through the ISPN cluster, right? Other local threads trying to access files in the directory would wait on the lock owner, and threads in other parts of the cluster should do likewise, shouldn't they?